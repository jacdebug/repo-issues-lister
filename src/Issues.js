import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { compose, renderComponent, lifecycle, branch } from "recompose";
import { selectIssues } from "./reducers";
import { loadIssues } from "./actions";
import Loading from "./Loading";

export const Issues = ({ issues }) => (
  <ul>{issues.map(({ id, title }) => <li key={id}>{title}</li>)}</ul>
);

Issues.propTypes = {
  issues: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      title: PropTypes.string.isRequired
    })
  ).isRequired
};

export default compose(
  connect(
    (state, props) => ({
      issues: selectIssues(state, props.repo)
    }),
    dispatch => ({
      loadIssues: repo => dispatch(loadIssues(repo))
    })
  ),
  lifecycle({
    componentDidMount() {
      if (!this.props.issues) {
        this.props.loadIssues(this.props.repo);
      }
    }
  }),
  branch(props => !props.issues, renderComponent(() => <Loading />))
)(Issues);
