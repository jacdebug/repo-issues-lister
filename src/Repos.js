import React, { Fragment } from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { compose, renderComponent, lifecycle, branch } from "recompose";
import { selectRepos, selectShowIssueFor } from "./reducers";
import { loadRepos, showIssueFor } from "./actions";
import Loading from "./Loading";
import Issues from "./Issues";

export const Repos = ({ repos, onClickRepo, showIssueFor }) => (
  <ul>
    {repos.map(({ id, name }) => (
      <Fragment key={id}>
        <li onClick={() => onClickRepo(name)}>{name}</li>
        {showIssueFor === name && <Issues repo={name} />}
      </Fragment>
    ))}
  </ul>
);

Repos.propTypes = {
  repos: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired
    })
  ).isRequired,
  onClickRepo: PropTypes.func.isRequired,
  showIssueFor: PropTypes.string.isRequired
};

export default compose(
  connect(
    state => ({
      repos: selectRepos(state),
      showIssueFor: selectShowIssueFor(state)
    }),
    dispatch => ({
      onClickRepo: repo => dispatch(showIssueFor(repo)),
      loadRepos: () => dispatch(loadRepos())
    })
  ),
  lifecycle({
    componentDidMount() {
      if (!this.props.repos.length) {
        this.props.loadRepos();
      }
    }
  }),
  branch(props => !props.repos.length, renderComponent(() => <Loading />))
)(Repos);
