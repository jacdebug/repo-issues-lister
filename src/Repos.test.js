import React from 'react';
import { shallow } from 'enzyme';
import { Repos } from './Repos';
import Issues from './Issues';

const noop = () => {};
const repos = () => [
  {
    id: 1,
    name: 'repo one name',
  },
  {
    id: 2,
    name: 'repo two name',
  },
];

describe('<Repos />', () => {
  it('should render correctly', () => {
    const component = shallow(
      <Repos repos={repos()} onClickRepo={noop} showIssueFor="" />
    );

    expect(component).toMatchSnapshot();
    expect(component.find('ul').length).toBe(1);
    expect(component.find('li').length).toBe(2);
  });

  it('should render Issues as per showIssueFor', () => {
    const component = shallow(
      <Repos repos={repos()} onClickRepo={noop} showIssueFor="repo two name" />
    );

    expect(component).toMatchSnapshot();
    expect(component.find('ul').length).toBe(1);
    expect(component.find('li').length).toBe(2);
    expect(component.find(Issues).length).toBe(1);
  });

  it('should call onClickRepo', () => {
    const mockOnClickRepo = jest.fn();
    const component = shallow(
      <Repos
        repos={repos()}
        onClickRepo={mockOnClickRepo}
        showIssueFor="repo two name"
      />
    );
    component
      .find('li')
      .at(1)
      .prop('onClick')();

    expect(mockOnClickRepo).toBeCalledWith('repo two name');
  });
});
