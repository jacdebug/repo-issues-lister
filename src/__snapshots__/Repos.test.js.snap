// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Repos /> should render Issues as per showIssueFor 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    onClickRepo={[Function]}
    repos={
        Array [
            Object {
              "id": 1,
              "name": "repo one name",
            },
            Object {
              "id": 2,
              "name": "repo two name",
            },
          ]
    }
    showIssueFor="repo two name"
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Unknown>
          <li
                    onClick={[Function]}
          >
                    repo one name
          </li>
</Unknown>,
        <Unknown>
          <li
                    onClick={[Function]}
          >
                    repo two name
          </li>
          <Connect(lifecycle(branch(Component)))
                    repo="repo two name"
          />
</Unknown>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <li
              onClick={[Function]}
>
              repo one name
</li>,
            false,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "repo one name",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "repo one name",
            "type": "li",
          },
          false,
        ],
        "type": Symbol(react.fragment),
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <li
              onClick={[Function]}
>
              repo two name
</li>,
            <Connect(lifecycle(branch(Component)))
              repo="repo two name"
/>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "repo two name",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "repo two name",
            "type": "li",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "repo": "repo two name",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": Symbol(react.fragment),
      },
    ],
    "type": "ul",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Unknown>
            <li
                        onClick={[Function]}
            >
                        repo one name
            </li>
</Unknown>,
          <Unknown>
            <li
                        onClick={[Function]}
            >
                        repo two name
            </li>
            <Connect(lifecycle(branch(Component)))
                        repo="repo two name"
            />
</Unknown>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <li
                onClick={[Function]}
>
                repo one name
</li>,
              false,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "repo one name",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "repo one name",
              "type": "li",
            },
            false,
          ],
          "type": Symbol(react.fragment),
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <li
                onClick={[Function]}
>
                repo two name
</li>,
              <Connect(lifecycle(branch(Component)))
                repo="repo two name"
/>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "repo two name",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "repo two name",
              "type": "li",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "repo": "repo two name",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": Symbol(react.fragment),
        },
      ],
      "type": "ul",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`<Repos /> should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown
    onClickRepo={[Function]}
    repos={
        Array [
            Object {
              "id": 1,
              "name": "repo one name",
            },
            Object {
              "id": 2,
              "name": "repo two name",
            },
          ]
    }
    showIssueFor=""
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Unknown>
          <li
                    onClick={[Function]}
          >
                    repo one name
          </li>
</Unknown>,
        <Unknown>
          <li
                    onClick={[Function]}
          >
                    repo two name
          </li>
</Unknown>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <li
              onClick={[Function]}
>
              repo one name
</li>,
            false,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "repo one name",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "repo one name",
            "type": "li",
          },
          false,
        ],
        "type": Symbol(react.fragment),
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <li
              onClick={[Function]}
>
              repo two name
</li>,
            false,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "repo two name",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": "repo two name",
            "type": "li",
          },
          false,
        ],
        "type": Symbol(react.fragment),
      },
    ],
    "type": "ul",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Unknown>
            <li
                        onClick={[Function]}
            >
                        repo one name
            </li>
</Unknown>,
          <Unknown>
            <li
                        onClick={[Function]}
            >
                        repo two name
            </li>
</Unknown>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <li
                onClick={[Function]}
>
                repo one name
</li>,
              false,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "repo one name",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "repo one name",
              "type": "li",
            },
            false,
          ],
          "type": Symbol(react.fragment),
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <li
                onClick={[Function]}
>
                repo two name
</li>,
              false,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "repo two name",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": "repo two name",
              "type": "li",
            },
            false,
          ],
          "type": Symbol(react.fragment),
        },
      ],
      "type": "ul",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
